### PR TITLE
117 implement deezer service

### LINKED ISSUE #117: Implement Deezer Service
**Short description:**
Add the service `Deezer`

**Describe the solution you'd like**
Connect via OAuth2 to a Deezer account, then get some actions from the service and maybe some reactions.
Add a "Connect to Deezer" button in the `Auth menu`

### PR DESCRIPTION
**Pull request by tbellicha**
# Purpose

The purpose of this pull request is to add the service Deezer.

Fixes #117

## Added features

* This service has been added to the application. It includes:
  * A reaction which does create a new empty playlist.
  * A reaction which does clear a playlist.
  * A reaction which does delete a playlist.
  * A reaction which does add few tracks from history to a playlist.
  * A reaction which does add few tracks from recommendations to a playlist.

### CODE REVIEW

File: server/services/deezer/common/add_tracks_to_playlist.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+'use strict'
+
+const axios = require('axios')
+
+async function addTracksToPlaylist (playlistId, tracksId, deezerToken) {
```
Comment (LucasTesnier) [inline server/services/deezer/common/add_tracks_to_playlist.js:12]: Missing doc

File: server/services/deezer/getters/user_playlists.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+'use strict'
+
+const axios = require('axios')
+
+async function getUserPlaylists (deezerId, deezerToken) {
```
Comment (LucasTesnier) [inline server/services/deezer/getters/user_playlists.js:11]: Missing doc

File: server/services/deezer/init.js [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+ */
+const createDeezerService = async () => {
+  try {
+    const deezer = await database.prisma.Service.create({
+      data: {
+        name: 'Deezer',
+        description: 'Deezer service',
+        primaryColor: '#323232',
+        secondaryColor: '#EF5466',
+        icon: '',
```
Comment (LucasTesnier) [inline server/services/deezer/init.js]: Where's the icon ?

File: server/services/deezer/init.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        isEnable: true,
+        Actions: {
```
Comment (LucasTesnier) [inline server/services/deezer/init.js]: If there's now action please do not add a false one

File: server/services/deezer/init.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+                    GetterUrl: '/api/services/deezer/getUserPlaylists'
+                  }
+                ]
+              }
+            },
+            {
+              name: 'addHistoryToPlaylist',
+              code: 'DZR-04',
+              description:
+                'Add a few tracks of your history to a owned playlist',
```
Comment (LucasTesnier) [inline server/services/deezer/init.js:72]: Maybe an X param ? For "few"

File: server/services/deezer/reactions/add_history_to_playlist.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
+'use strict'
+
+const axios = require('axios')
+
+const { replaceDynamicParameters } = require('../../glue/dynamic_parameters.js')
+const { addTracksToPlaylist } = require('../common/add_tracks_to_playlist.js')
+
+async function deezerAddHistoryToPlaylistFromArea (Area, dynamicParameters) {
```
Comment (LucasTesnier) [inline server/services/deezer/reactions/add_history_to_playlist.js:14]: Doc

File: server/services/deezer/reactions/add_history_to_playlist.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  const reactionParameters = Area.ReactionParameters
+  let playlistId = reactionParameters.find(
+    parameter => parameter.Parameter.name == 'playlistId'
+  ).value
+  playlistId = replaceDynamicParameters(playlistId, dynamicParameters)
+  return addHistoryToPlaylist(playlistId, Area.User.deezerId, Area.User.deezerToken)
+}
+
+async function addHistoryToPlaylist (playlistId, deezerId, deezerToken) {
```
Comment (LucasTesnier) [inline server/services/deezer/reactions/add_history_to_playlist.js]: Doc

File: server/services/deezer/reactions/add_recommendations_to_playlist.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
+'use strict'
+
+const axios = require('axios')
+
+const { replaceDynamicParameters } = require('../../glue/dynamic_parameters.js')
+const { addTracksToPlaylist } = require('../common/add_tracks_to_playlist.js')
+
+async function deezerAddRecommendationsToPlaylistFromArea (
```
Comment (LucasTesnier) [inline server/services/deezer/reactions/add_recommendations_to_playlist.js:14]: Doc

File: server/services/deezer/reactions/add_recommendations_to_playlist.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  ).value
+  playlistId = replaceDynamicParameters(playlistId, dynamicParameters)
+  return addRecommendationsToPlaylist(
+    playlistId,
+    Area.User.deezerId,
+    Area.User.deezerToken
+  )
+}
+
+async function addRecommendationsToPlaylist (playlistId, deezerId, deezerToken) {
```
Comment (LucasTesnier) [inline server/services/deezer/reactions/add_recommendations_to_playlist.js]: Doc

File: server/services/deezer/reactions/clear_playlist.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
+'use strict'
+
+const axios = require('axios')
+
+const { replaceDynamicParameters } = require('../../glue/dynamic_parameters.js')
+
```
Comment (LucasTesnier) [inline server/services/deezer/reactions/clear_playlist.js:6]: Doc

File: server/services/deezer/reactions/clear_playlist.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+function deezerClearPlaylistFromArea (Area, dynamicParameters) {
+  const reactionParameters = Area.ReactionParameters
+  let playlistId = reactionParameters.find(
+    parameter => parameter.Parameter.name == 'playlistId'
+  ).value
+  playlistId = replaceDynamicParameters(playlistId, dynamicParameters)
+  return clearPlaylist(playlistId, Area.User.deezerToken)
+}
+
```
Comment (LucasTesnier) [inline server/services/deezer/reactions/clear_playlist.js:21]: Doc

File: server/services/deezer/reactions/create_playlist.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
+'use strict'
+
+const axios = require('axios')
+
+const { replaceDynamicParameters } = require('../../glue/dynamic_parameters.js')
+
+function deezerCreatePlaylistFromArea (Area, dynamicParameters) {
```
Comment (LucasTesnier) [inline server/services/deezer/reactions/create_playlist.js:13]: Doc

File: server/services/deezer/reactions/create_playlist.js [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+  const reactionParameters = Area.ReactionParameters
+  let title = reactionParameters.find(
+    parameter => parameter.Parameter.name == 'title'
+  ).value
+  title = replaceDynamicParameters(title, dynamicParameters)
+  return createPlaylist(title, Area.User.deezerId, Area.User.deezerToken)
+}
+
+async function createPlaylist (title, deezerId, deezerToken) {
```
Comment (LucasTesnier) [inline server/services/deezer/reactions/create_playlist.js:29]: Doc

File: server/services/deezer/reactions/delete_playlist.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
+'use strict'
+
+const axios = require('axios')
+
+const { replaceDynamicParameters } = require('../../glue/dynamic_parameters.js')
+
+function deezerDeletePlaylistFromArea (Area, dynamicParameters) {
```
Comment (LucasTesnier) [inline server/services/deezer/reactions/delete_playlist.js:13]: Doc

File: server/services/deezer/reactions/delete_playlist.js [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+  const reactionParameters = Area.ReactionParameters
+  let playlistId = reactionParameters.find(
+    parameter => parameter.Parameter.name == 'playlistId'
+  ).value
+  playlistId = replaceDynamicParameters(playlistId, dynamicParameters)
+  return deletePlaylist(playlistId, Area.User.deezerToken)
+}
+
+async function deletePlaylist (playlistId, deezerToken) {
```
Comment (LucasTesnier) [inline server/services/deezer/reactions/delete_playlist.js:28]: Doc

**Review by LucasTesnier — CHANGES_REQUESTED** [review]
Answer request

File: server/server_app.js [hunk 16]
```diff
@@ -880,6 +944,39 @@ app.post('/api/dev/user/create', async (req, res) => {
   }
 })
 
+/**
+ * @brief Add the Deezer ID to the user in the database.
+ */
```
Comment (ZQUEMA) [inline server/server_app.js:961]: complete the documentation with the needed body fields

File: server/services/deezer/init.js [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+                    GetterUrl: '/api/services/deezer/getUserPlaylists'
+                  }
+                ]
+              }
+            },
+            {
+              name: 'addRecommendationsToPlaylist',
+              code: 'DZR-05',
+              description:
+                'Add a few tracks of your recommendations to a owned playlist',
```
Comment (ZQUEMA) [inline server/services/deezer/init.js:94]: Same as above

**Review by ZQUEMA — CHANGES_REQUESTED** [review]
Some needed changes

File: server/server_app.js [hunk 18]
```diff
@@ -808,6 +848,30 @@ app.get(
   }
 )
 
+app.get(
```
Comment (LucasTesnier) [inline server/server_app.js:860]: missing doc

**Review by LucasTesnier — CHANGES_REQUESTED** [review]
doc

**Review by TomDUVAL-MAHE — DISMISSED** [review]

**Review by TomDUVAL-MAHE — DISMISSED** [review]

**Review by LucasTesnier — DISMISSED** [review]

**Review by ZQUEMA — APPROVED** [review]

**Review by LucasTesnier — APPROVED** [review]

**Pull request merged**